File: ts/twoSum2.ts
function twoSum(nums: number[], target: number): number[] {
  const result: number[] = [];
  const hashMap = new Map<number, number>(); // { value: index }

  for (let i = 0; i < nums.length; i++) {
    const diff = target - nums[i];

    if (hashMap.has(diff)) {
      result.push(hashMap.get(diff) as number, i);
      break;
    }

    hashMap.set(nums[i], i);
  }

  return result;
}

export default () => {
  const nums = [-3, 4, 3, 90];
  const target = 0;
  // const nums = [0, 4, 3, 0];
  // const target = 0;
  // const nums = [3, 3];
  // const target = 6;
  // const nums = [3, 2, 4];
  // const target = 6;
  // const nums = [2, 7, 11, 15];
  // const target = 9;

  const result = twoSum(nums, target);
  console.log(result);
};

File: rs/src/two_sum2.rs
use std::collections::HashMap;

fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut map: HashMap<i32, i32> = HashMap::new();
    let mut result: Vec<i32> = vec![];

    for (i, &n) in nums.iter().enumerate() {
        let diff = target - n;

        if let Some(&val) = map.get(&diff) {
            result.push(val);
            result.push(i as i32);

            break;
        } else {
            map.insert(n, i as i32);
        }
    }

    return result;
}

pub fn main() {
    let nums = vec![2, 7, 11, 15];
    let target: i32 = 9;

    let result = two_sum(nums, target);
    println!("{:?}", result);
}
